Create 2021-06-09-solidity-0.8.5-release-announcement.md (#141)
ethereum/solidity-blog: _posts/2021-06-09-solidity-0.8.5-release-announcement.md

---
layout: post
published: true
title: 'Solidity 0.8.5 Release Announcement'
date: '2021-06-09'
author: Solidity Team
category: Releases
---

[Solidity v0.8.5](https://github.com/ethereum/solidity/releases/tag/v0.8.5) allows byte conversion, adds ``verbatim``builtin function to inject arbitrary bytecode in Yul and fixes several smaller bugs.

## Notable New Features
### Bytes Conversion

This release introduces ability to convert `bytes` and `bytes` slices to fixed bytes types `bytes1`/../`bytes32`.

> In case array is longer than the target fixed type, truncation at the end will happen:

Reviewer: Maybe we should add what part gets truncated?

```solidity
function f(bytes calldata c) public view returns (bytes8) {

> 	require(c.length == 16, "");

Reviewer: ```suggestion
	require(c.length == 16);
```

		// bytes4 sig = bytes4(_payload);
		if (sig == bytes4(keccak256("setOwner(address)"))) {
			address owner = abi.decode(_payload[4:], (address));
			require(owner != address(0), "Address of owner cannot be zero.");
		}
		(bool status,) = client.delegatecall(_payload);
		require(status, "Forwarded call failed.");
	}
}
```

More details can be found in
["documentation"](https://docs.soliditylang.org/en/v0.8.5/types.html#explicit-conversions)

### ``Verbatim``in Yul

This release introduces the set of `verbatim` builtin functions for Yul that allows you inject
arbitrary bytecode into the binary. This is currently only available via pure Yul, i.e., it is not
accessible via inline assembly.

The functions are `verbatim<n>i_<m>o("<data>", ...)`, where
- `n` is a decimal between 0 and 99 that specifies the number of input stack slots / variables,
- `m` is a decimal between 0 and 99 that specifies the number of output stack slots / variables,
- `data` is a string literal that contains the sequence of bytes.

For example,

[EIP-3198](https://eips.ethereum.org/EIPS/eip-3198) proposes the the opcode `BASEFEE` at `0x48`.

> Where BASEFEE is defined in [EIP-1559](https://eips.ethereum.org/EIPS/eip-3198). Since the Solidity

Reviewer: Rewrote it.

compiler currently does not support this opcode, one can use `verbatim` to implement this in Yul.

```Yul
{
	function basefee() -> out {
		out := verbatim_0i_1o(hex"48")
	}

	sstore(0, basefee())
}
```

> Here's another example that 

Author: Sentence incomplete?

Reviewer: Fixed.

```Yul
let x := calldataload(0)
// The hex"600202" corresponds to EVM instructions:
// PUSH 02 MUL
// That is, it multiplies x by 2.
let double := verbatim_1i_1o(hex"600202", x)
```

The above code will result in a `dup1` opcode to retrieve `x` (the optimizer may directly use the
result of the `calldataload` opcode, though) directly followed by `600202`. The code is assumed to
consume the (copied) value of `x` and produce the result on the top of the stack. The compiler then
generates code to allocate a stack slot for `double` and store the result there.

More details can be found in the

> [documentation](https://docs.soliditylang.org/en/v0.8.5/yul.html#verbatim)

Author: ```suggestion
[documentation](https://docs.soliditylang.org/en/v0.8.5/yul.html#verbatim).
```

---
layout: post
published: true
title: 'Solidity 0.8.5 Release Announcement'
date: '2021-06-09'
author: Solidity Team
category: Releases
---

> [Solidity v0.8.5](https://github.com/ethereum/solidity/releases/tag/v0.8.5) allows byte conversion, adds ``verbatim``builtin function to inject arbitrary bytecode in Yul and fixes several smaller bugs.

Reviewer 2: ```suggestion
[Solidity v0.8.5](https://github.com/ethereum/solidity/releases/tag/v0.8.5) allows converting `bytes` into `bytesNN` values, adds ``verbatim``builtin function to inject arbitrary bytecode in Yul and fixes several smaller bugs.
```

---
layout: post
published: true
title: 'Solidity 0.8.5 Release Announcement'
date: '2021-06-09'
author: Solidity Team
category: Releases
---

[Solidity v0.8.5](https://github.com/ethereum/solidity/releases/tag/v0.8.5) allows byte conversion, adds ``verbatim``builtin function to inject arbitrary bytecode in Yul and fixes several smaller bugs.

## Notable New Features
### Bytes Conversion

### ``Verbatim``in Yul

> 

Reviewer: ```suggestion

This release introduces the set of `verbatim` builtin functions for Yul that allows you inject
arbitrary bytecode into the binary. This is currently only available via pure Yul, i.e., it is not
accessible via inline assembly.

The functions are `verbatim<n>i_<m>o("<data>", ...)`, where
- `n` is a decimal between 0 and 99 that specifies the number of input stack slots / variables,
- `m` is a decimal between 0 and 99 that specifies the number of output stack slots / variables,
- `data` is a string literal that contains the sequence of bytes.

For example,

[EIP-3198](https://eips.ethereum.org/EIPS/eip-3198) proposes the the opcode `BASEFEE` at `0x48`.
Where BASEFEE is defined in [EIP-1559](https://eips.ethereum.org/EIPS/eip-3198). Since the Solidity
compiler currently does not support this opcode, one can use `verbatim` to implement this in Yul.

```Yul
{
	function basefee() -> out {
		out := verbatim_0i_1o(hex"48")
	}

	sstore(0, basefee())
}
```

Here's another example that 

```Yul
let x := calldataload(0)
// The hex"600202" corresponds to EVM instructions:
// PUSH 02 MUL
// That is, it multiplies x by 2.
let double := verbatim_1i_1o(hex"600202", x)
```

The above code will result in a `dup1` opcode to retrieve `x` (the optimizer may directly use the
result of the `calldataload` opcode, though) directly followed by `600202`. The code is assumed to
consume the (copied) value of `x` and produce the result on the top of the stack. The compiler then
generates code to allocate a stack slot for `double` and store the result there.

More details can be found in the [documentation](https://docs.soliditylang.org/en/v0.8.5/yul.html#verbatim).

```

Reviewer: I'll just push this. The suggestion got messed up.
